### PR TITLE
Improve handling of incoming i64 syscall arguments.    

### PR DESCRIPTION
**Pull request by sbc100**
Add new `receiveI64ParamAsI53` to replace the old                           
`receiveI64ParamAsDouble`.  The new function checks the value fits          
within I53 and requires an `onError` value to be returned in the case of 
out-of-bounds.                                                                                               

`receiveI64ParamAsI53` makes use of the existing `convertI32PairToI53`   
utility.                                                                 

There is now some amount of duplication between `test_parseTools`        
and `js_library_i64_params` which I would prefer to address as           
a followup.

### CODE REVIEW

File: src/library_int53.js
Comment (dschuff) [inline src/library_int53.js:119]: I think this would probably be better named "clamping" or something rather than "signaling". I think of "signaling" as more like throwing (e.g. like a signaling NaN).

Actually I guess "clamping" wouldn't be right either since that would imply returning the max value rather than a sentinel value. Not sure what a better idea would be.

**Review by dschuff — APPROVED** [review]

File: src/library_int53.js
Comment (sbc100) [reply src/library_int53.js:119]: Don't you think that returns NaN counts as a "signal"?   Maybe "Checked" would be a better adjective, but the other ones in this files end in `Signaling`.. but then they also do actually throw.

**Comment by sbc100** [comment]
Renamed to `Checked`

File: src/parseTools.js
Comment (juj) [inline src/parseTools.js:1109]: End users may also have been using `receiveI64ParamAsDouble` as they may need to convert I64 pairs to doubles as well? This change will break their code?

Comment (sbc100) [reply src/parseTools.js:1109]: I guess making the `onEerror` parameter optional and adding a test for that case would be address that concern?

File: src/library_int53.js
Comment (juj) [reply src/library_int53.js]: Maybe `convertI32PairToI53OrNaN`? The "I53" hints that it would return an integer, but then the function can also return a non-integer (NaN)

File: src/parseTools.js
Comment (juj) [reply src/parseTools.js:1109]: yeah, certainly, that would make the old form work as well.

**Review by kripken — COMMENTED** [review]
When would the `*Checked` versions be used as opposed to the unchecked? Do we want to gate that on ASSERTIONS maybe?

**Comment by sbc100** [comment]
The usage in `receiveI64ParamAsI53` is always checked since we want to be able to return an error code when the argument is out of range.   We leave it up to the caller to decide if they want to assert, or return an error the caller.   For system call argument checking I think returning a error code seems like the right thing to do.

File: src/parseTools.js
Comment (sbc100) [reply src/parseTools.js]: I ended up going with a completely new function `receiveI64ParamAsI53` rather and deprecating the old/unused `receiveI64ParamAsDouble` .    This avoids over-complicating the new code with unused/legacy behaviour and gives the functions names more semantic meaning (easier to grep for usage, for example).

File: src/library_int53.js
Comment (sbc100) [reply src/library_int53.js]: I prefer `convertI32PairToI53Checked` myself since it tells me what the function is doing..  but since its a mostly-internal-facing function is don't feel strongly.. I will rename if you do.

Comment (juj) [reply src/library_int53.js]: > since its a mostly-internal-facing function

The API in library_int53.js is inteded to be a public API for end users. It is common that end users also need to do i64/i53/i32 pair/Number marshalling across the boundary, so it is good that the API that the system libraries use would be the same that is exposed to end users.

Authored PRs #16865 and #16866 to implement a lean `$convertI32PairToI53Checked()` that is tested using the same machinery that the rest of library_int53.js is. Maybe that can be helpful?

Comment (juj) [reply src/library_int53.js]: > I prefer convertI32PairToI53Checked myself since it tells me what the function is doing

I find it is odd to have the name "ToI53" there, when the result might not be an integer, and the name "checked" seems a bit ambiguous as to what kind of check is performed, and what the result of that check will be. Though maybe the "OrNaN" name would not either tell when exactly the NaN would appear, so guess that name is not much better either. Proposed the `Checked` name in the above PR.

File: src/library_int53.js
Comment (juj) [inline src/library_int53.js:98]: Having a NaN-returning form of the function might not obviate the need for a signaling (exception throwing) variant of the function, so this TODO would still feel relevant even after a Checked version is added?

File: src/library_int53.js
Comment (juj) [inline src/library_int53.js:132]: Curiously, the values of `MIN_SAFE_INTEGER` and `MAX_SAFE_INTEGER` are actually off-by-one for this purpose for deciding whether a U64 or a I64 can safely be converted to a Number.

**Comment by kripken** [comment]
> The usage in receiveI64ParamAsI53 is always checked since we want to be able to return an error code when the argument is out of range. We leave it up to the caller to decide if they want to assert, or return an error the caller.

Is there an example of the assertion case?

I am guessing for an assert do you mean to replace the last param, the return value, with an assert? That would still contain the checking code even if assertions are disabled, unless I'm not seeing how it's removed?

**Comment by sbc100** [comment]
> > The usage in receiveI64ParamAsI53 is always checked since we want to be able to return an error code when the argument is out of range. We leave it up to the caller to decide if they want to assert, or return an error the caller.
> 
> Is there an example of the assertion case?
> 
> I am guessing for an assert do you mean to replace the last param, the return value, with an assert? That would still contain the checking code even if assertions are disabled, unless I'm not seeing how it's removed?

Sorry I meant 'throw' not assert.  We have three different types of functions in this file:

1. No checks
2. Check and throw on failure  (Signaling)
3. Check and return error on failure (Checked)

As far as I know all our internal use cases (at least the system layer) want (3) so we can return an error code to the native side.

**Comment by sbc100** [comment]
Waiting on  #16865 and #16866 to land

**Comment by kripken** [comment]
For syscalls I agree (3; return error code) makes sense. For (2; throw) I think we want that with assertions enabled, only? Or is there another situation?

**Comment by sbc100** [comment]
> For syscalls I agree (3; return error code) makes sense. For (2; throw) I think we want that with assertions enabled, only? Or is there another situation?

Some optimized builds might also prefer (1) if it reduces code size or improves performance.  i.e. most program know that their size_t and off_t arguments are going to be in range so might not want the runtime checks on every syscall.

**Comment by sbc100** [comment]
@juj are you ready to land #16865 and #16866 or can then land after this PR?

**Pull request merged**
